src/poetry/core/masonry/builders/builder.py

```diff
@@ -404,3 +420,12 @@ def relative_to_source_root(self) -> Path:
             return self.path.relative_to(self.source_root)
 
         return self.path
+
+    def concatenation_target_dir_relative_to_source_root(self) -> Path:
+        if self.source_root is not None and self.target_dir is not None:
+            return self.target_dir / self.path.relative_to(self.source_root)
+
+        elif self.source_root is not None:
+            return self.path.relative_to(self.source_root)
+
+        return self.path
```

Comment:
```suggestion
        path = self.relative_to_source_root()
        if self.source_root is not None and self.target_dir is not None:
            return self.target_dir / path
        return path
```

And two more things:

- IMO, the method's name is a bit clunky. May we call it `relative_to_target_root()` because that's what it is in the end, isn't it?
- If `source_root` is None and `target_dir` is not None, then the result does not make sense. If I don't miss anything, `source_root` is always passed and will never be None. Maybe, we should just make it mandatory?